test/helpers/position_assertions.cc

```diff
@@ -1129,4 +1130,119 @@ void ApplyCodeActionAssertion::check(const UnorderedMap<std::string, std::shared
+    if (auto symbolInfos = get_if<vector<unique_ptr<SymbolInformation>>>(
+            &get<variant<JSONNullObject, vector<unique_ptr<SymbolInformation>>>>(result))) {
+        for (auto &symbolInfo : *symbolInfos) {
+            symbols.push_back(shared_ptr<SymbolInformation>(std::move(symbolInfo)));
+        }
+    }
+    std::map<std::string, vector<shared_ptr<SymbolInformation>>> symbolsByLine;
+    for (auto &symbol : symbols) {
+        auto &location = symbol->location;
+        auto line = fmt::format("{}:{:{}}", location->uri, location->range->start->line, lineWidth);
+        symbolsByLine[line].push_back(symbol);
+    }
+
+    bool success = true;
+    // Show any expected symbols that weren't returned by LSP
+    for (auto entry : assertionsByLine) {
+        auto lineInfo = entry.first;
+        auto expected = entry.second;
+        auto &actual = symbolsByLine[lineInfo];
+        if (expected.size() == actual.size()) {
+            continue;
+        }
+        success = false;
+        auto assertion = expected[0];
+        auto location = assertion->getLocation(uriPrefix);
+        auto sourceLine = getLine(sourceFileContents, uriPrefix, *location.get());
+        ADD_FAILURE_AT(location->uri.c_str(), assertion->range->start->line + 1)
+            << fmt::format("WorkspaceSymbol query `{}` did not return a reference to:\n{}\n", query,
+                           prettyPrintRangeComment(sourceLine, *assertion->range, ""));
+    }
+
+    // Show any symbols returned by LSP that weren't asserted
+    for (auto entry : symbolsByLine) {
+        auto lineInfo = entry.first;
+        auto actual = entry.second;
+        auto &expected = assertionsByLine[lineInfo];
+        if (expected.size() == actual.size()) {
+            continue;
+        }
+        success = false;
+        auto symbolInfo = actual[0];
+        auto &location = symbolInfo->location;
+        auto sourceLine = getLine(sourceFileContents, uriPrefix, *location.get());
+        ADD_FAILURE_AT(location->uri.c_str(), location->range->start->line + 1)
+            << fmt::format("WorkspaceSymbol query `{}` returned an unexpected reference to:\n{}\n", query,
+                           prettyPrintRangeComment(sourceLine, *location->range, ""));
+    }
+
+    ASSERT_TRUE(success) << fmt::format("For query \"{}\"", query);
+}
+
+void SymbolSearchAssertion::checkAll(const vector<shared_ptr<RangeAssertion>> &assertions,
+                                     const UnorderedMap<string, shared_ptr<core::File>> &sourceFileContents,
+                                     LSPWrapper &lspWrapper, int &nextId, string_view uriPrefix, string errorPrefix) {
+    UnorderedMap<std::string, vector<shared_ptr<SymbolSearchAssertion>>> queryToAssertionsMap;
+    for (auto &assertion : assertions) {
+        if (auto searchAssertion = dynamic_pointer_cast<SymbolSearchAssertion>(assertion)) {
+            queryToAssertionsMap[searchAssertion->query].push_back(searchAssertion);
```

Comment:
One bit of magic here that we didn't discuss: This works because `queryToAssertion["somestring"]` will auto-create a vector for key `"somestring"` the first time you access it.

(I'm not requesting that you change anything, I just wanted to explain this one tricky C++ thing that you might not have known.)